python/lsst/cp/pipe/ptc/cpPtcSolve.py

```diff
@@ -1077,14 +1110,168 @@ def _makeZeroSafe(self, array, substituteValue=1e-9):
+                       f"in initialExpIdMask: {initialExpIdMask} are bad."
+                       f"Setting {ampName} to BAD.")
+                self.log.warning(msg)
+                # Fill entries with NaNs
+                self.fillBadAmp(dataset, ampName)
+                continue
+
+            mask = goodPoints.copy()
+
+            # Compute the extended mask
+            preTurnoff = dataset.ptcTurnoff[ampName]
+            turnoffIdx = np.argwhere(meanVecSorted == preTurnoff)[0][0]
+            varianceAtTurnoff = varVecSorted[turnoffIdx]
+
+            # Add points up to some threshold below the variance of the turnoff
+            pointsToFit = (meanVecSorted >= preTurnoff) * \
+                          (meanVecSorted < preTurnoff+self.config.extendRollofSearchMaskSizeAdu) * \
+                          (varVecSorted <= varianceAtTurnoff)
+
+            # Retain the original mask below the turnoff
+            pointsToFit = np.logical_or(mask, pointsToFit)
+            if np.count_nonzero(pointsToFit) == np.count_nonzero(mask):
+                self.log.warning("Expanding fit to include saturation, but no points detected above "
+                                 f"initial computed PTC turnoff for amp {ampName}.")
+                continue
+
+            # Fit initialization
+            ptcFunc = funcAstierWithRolloff
+            parsIniPtc = dataset.ptcFitPars[ampName]  # a00, gain, noise^2
+
+            # Estimate initial parameters
+            estimateTurnoff = preTurnoff
+            estimateTau = -1200  # This turns out to be a generally good guess!
+            parsIniPtc = np.append(
+                parsIniPtc,
+                [estimateTurnoff, estimateTau],
+            )
+
+            # Set initial bounds
+            bounds = self._boundsForAstier(
+                parsIniPtc,
+                lowers=[-1e-4, 0.1, 0, min(preTurnoff-20000, 0), -10000],
+                uppers=[0, 10.0, 2000, preTurnoff, -100],
+            )
+
+            # Perform the fit
+            res = least_squares(
+                errFunc,
+                parsIniPtc,
+                bounds=bounds,
+                args=(meanVecSorted[pointsToFit], varVecSorted[pointsToFit]),
+            )
+            pars = res.x
+            originalModelPars = pars[:-2]
+            J = res.jac
+            cov = np.linalg.inv(J.T @ J)
+            parErrors = np.sqrt(np.diag(cov))
+            a00, gain, noiseSquared, rolloff, tau = pars
+            a00Err, gainErr, noiseSquaredError, rolloffError, tauError = parErrors
+
+            if not res.success:
+                self.log.warning(
+                    "Fit with saturation roll off model did not succeed for "
+                    f"amp {ampName}."
+                )
+                continue
+            else:
+                self.log.info("Fit with saturation roll off model returned estimates: "
+                              f"rolloff={rolloff} adu ({rolloff * gain} el) and tau={tau}")
```

Comment:
Unfortunately, f-strings are not technically allowed in our logs.  We have to do the `%f` formatting.